EDM Diffusion Models

mnabian opened the pull request:
<!-- markdownlint-disable MD013-->
# Modulus Pull Request

## Description
<!-- Provide a standalone description of changes in this PR. -->
<!-- Reference any issues closed by this PR with "closes #1234". -->
<!-- Note: The pull request title will be included in the CHANGELOG. -->

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

## Dependencies

<!-- Call out any new dependencies needed if any -->

mnabian commented:
/blossom-ci

mnabian commented:
/blossom-ci

mnabian commented:
/blossom-ci

mnabian commented:
/blossom-ci

mnabian commented:
/blossom-ci

mnabian commented:
/blossom-ci

mnabian commented:
/blossom-ci

mnabian commented:
/blossom-ci

mnabian commented:
/blossom-ci

dallasfoster commented on .pre-commit-config.yaml:
```diff
@@ -49,6 +49,7 @@ repos:
     hooks:
     - id: ruff
       args: [--fix]
+      exclude: 'examples'
```
1. Is this exclusion of the examples folder going to be a general rule or is it because you would rather not have to lint the files in the present folder right now? 

2. Please keep exceptions and exclusions to the `pyproject.toml` file under the `[tool.ruff]` header.

dallasfoster commented on examples/generative/diffusion/conf/config.yaml:
```diff
@@ -0,0 +1,96 @@
+
+
+
+# # Weather-related
+# data_config: ?          # String to include the data config
+# task: ?                 # String to include the task
+# data_type: ?            # String to include the data type
+
+# # Regression
+# ckpt_unet: ?            # Checkpoint for the UNet to predict the mean
```
Unless you have a good reason for the inclusion of these commented lines, I would just as soon not include them in order to prevent confusion.

dallasfoster commented on examples/generative/diffusion/fid.py:
```diff
@@ -0,0 +1,202 @@
+    return mu.cpu().numpy(), sigma.cpu().numpy()
+
+
+def calculate_fid_from_inception_stats(mu, sigma, mu_ref, sigma_ref):
+    m = np.square(mu - mu_ref).sum()
+    s, _ = scipy.linalg.sqrtm(np.dot(sigma, sigma_ref), disp=False)
+    fid = m + np.trace(sigma + sigma_ref - s * 2)
+    return float(np.real(fid))
```
1. This should be moved to `modulus/metrics`.
2. scipy.ilnalg.sqrtm for this matrix is not stable generally.

dallasfoster commented on examples/generative/diffusion/generate.py:
```diff
@@ -0,0 +1,334 @@
+from modulus.launch.logging import PythonLogger, RankZeroLoggingWrapper
+
+
+def sampler(
```
This sampler is overly complex, poorly documented, and not modular. If we intend to support generative models like these in the future, it would behoove us to have general purpose langevin samplers.

mnabian replied:
Added to this epic: https://github.com/NVIDIA/modulus/issues/196

dallasfoster commented on modulus/metrics/diffusion/loss.py:
```diff
@@ -0,0 +1,376 @@
+
+    Note:
+    -----
+    Reference: Song, Y., Sohl-Dickstein, J., Kingma, D.P., Kumar, A., Ermon, S. and
```
I don't think it should be necessary for a user to reference the paper in order to understand the notation, terminology, and purpose of the class.  In this case, I think __call__ and sigma should have documentation as well.

mnabian replied:
Added docstrings for call. Sigma already has docstrings

dallasfoster commented on modulus/metrics/diffusion/loss.py:
```diff
@@ -0,0 +1,376 @@
+    equations. arXiv preprint arXiv:2011.13456.
+    """
+
+    def __init__(self, sigma_min: float = 0.02, sigma_max: float = 100.0):
```
Same comments as for `VPLoss`, documentation needs to be improved.

mnabian replied:
Added documentation for the `__call__` method

dallasfoster commented on modulus/metrics/diffusion/loss.py:
```diff
@@ -0,0 +1,376 @@
+        return loss
+
+
+# class EDMLoss:
```
Why is this class commented out? We should either uncomment or not include it in the code.

mnabian replied:
The `EDMLoss` class is WIP, to generalize it to different models including ResDiff. the commented out code helps to guide this generalization.

dallasfoster commented on modulus/metrics/diffusion/loss.py:
```diff
@@ -0,0 +1,376 @@
+        self.sigma_data = sigma_data
+
+        with torch.no_grad():
+            # resume_state_dump = '/lustre/fsw/nvresearch/mmardani/output/logs/edm_era5_cwb_12chans_fcn_4x_crop448_zarr_fulldata_unetregression/complex-mayfly_2023.04.25_16.42/output/training-state-006924.pt'
+            resume_state_dump = "/lustre/fsw/nvresearch/mmardani/output/logs/edm_era5_cwb_12chans_fcn_4x_crop448_zarr_fulldata_unetregression/mindful-parakeet_2023.04.03_22.54/output/training-state-042650.pt"
```
This is inappropriate to include in public code.

dallasfoster commented on modulus/metrics/diffusion/loss.py:
All of these classes should have better documentation but also consider whether they should be classes or not. Sometimes the class structure for metrics is just duct tape for poor workflows.

mnabian replied:
Improved the documentation. Added the class vs function issue to this epic: https://github.com/NVIDIA/modulus/issues/196

dallasfoster commented on modulus/models/diffusion/__init__.py:
Consider whether these networks should live under `models/diffusion`, `models/` or maybe even `models/unet`.

mnabian replied:
Added to this epic: https://github.com/NVIDIA/modulus/issues/196

dallasfoster commented on modulus/models/diffusion/utils.py:
It is not clear why this is necessary or why it mixes numpy and torch.

mnabian replied:
Added to this epic: https://github.com/NVIDIA/modulus/issues/196
I think this was mainly because they did not want to have an `nn` module as the input to their model, as that can potentially cause issues with pickling. I'll try to figure this out and most likely we should be able to get rid of this entirely.

dallasfoster commented on modulus/models/diffusion/layers.py:
```diff
@@ -0,0 +1,541 @@
+from modulus.models.diffusion import weight_init
+
+
+class Linear(torch.nn.Module):
```
Is this function necessary as a standalone layer under `models/diffusion`. It seems like the only purpose for its existance is to support layer initialization. If so, then maybe we should include it under `models/layers`.

mnabian replied:
Added to this epic: https://github.com/NVIDIA/modulus/issues/196

dallasfoster commented on modulus/models/diffusion/layers.py:
Consider whether these layers should be general enough to include under `models/layers`

mnabian replied:
Added to this epic: https://github.com/NVIDIA/modulus/issues/196

dallasfoster commented on examples/generative/diffusion/dataset/__init__.py:
The dataset function should really be reconsidered and determine whether they can be generalized and moved out of examples.

dallasfoster reviewed (commented):
There are a lot of places where documentation can be improved and notation is not entirely explain. 

The examples folder should at least include a readme with an explanation about how to use the example.

The placement of the layers and folders should be reconsidered and generalized wherever possible.

Creation of a generic, modular, langevin diffusion class should be considered as a separate epic.

There are several places with large chunks of commented out code, please consider uncommenting or removing these pieces of code.

Please let me know which of these and the other comments you think should be in-scope for this PR.

mnabian commented:
Thanks @dallasfoster for reviewing this PR!

> There are a lot of places where documentation can be improved and notation is not entirely explain.
I agree. I added more documentation to `loss.py`, and I will keep improving the documentation in the subsequent PRs.

> The examples folder should at least include a readme with an explanation about how to use the example.
The example is a work in progress. Once the example is ready, I will also include a README with detailed instructions.

> The placement of the layers and folders should be reconsidered and generalized wherever possible.
I agree. I have added this to the list of issues in this epic: https://github.com/NVIDIA/modulus/issues/196. Will decide about this once I develop the example and get a better understanding of what can be generalized.

> Creation of a generic, modular, langevin diffusion class should be considered as a separate epic.
For now I added it to this epic: https://github.com/NVIDIA/modulus/issues/196

> There are several places with large chunks of commented out code, please consider uncommenting or removing these pieces of code.
The commented out code is mostly the ResDiff code and it helps me generalize the existing code to add support for ResDiff and super-resolution task. I propose we keep them in this PR, and I can remove the remaining comments right before the release.

mnabian commented:
/blossom-ci

loliverhennigh commented on test/metrics/diffusion/test_losses.py:
```diff
@@ -0,0 +1,265 @@
+    assert isinstance(loss_value_with_augmentation, torch.Tensor)
+
+
+# RegressionLoss tests
```
Maybe delete this?

mnabian replied:
Thanks Oliver for your comment. Those tests are work in progress and I should be able to complete and uncomment those in a subsequent PR.

loliverhennigh reviewed (approved):
Hey @mnabian, to be honest I struggled a bit with giving this a review. I ran through all the code. There is quite a lot of detail with how then diffusion models work and I am not very familiar with all the variations in this space. Certainty a lot more complex then just basic regression like many of our examples.

At this point I don't really have any good advice about integrating this better with modulus other then some of the comments Dallas made such as moving over some of the layer functions over. I think we should continue to treat this as a long term project following your epic, https://github.com/NVIDIA/modulus/issues/196. Lets just get this in and as the examples build I think we will be able to work out abstractions that make sense...

mnabian commented:
/blossom-ci

dallasfoster reviewed (approved):
Thank you for adding the necessary tasks to the epic list. I also appreciate the additional documentation. I think at some point we would want to include the examples in the ruff linting, but I understand that might be a separate PR after the launch examples have been moved. Everything else here looks manageable.
